TopsortAnalytics/src/main/java/com/topsort/analytics/EventPipeline.kt

```diff
@@ -0,0 +1,173 @@
+
+val Context.eventDatastore: DataStore<Preferences> by preferencesDataStore(name = PREFERENCES_NAME)
+
+internal object EventPipeline {
+
+    private lateinit var applicationContext: Context
+
+    private val scope = CoroutineScope(SupervisorJob())
+    private val dispatcher = Dispatchers.IO
+
+    private val writeChannelImpressions = Channel<List<Impression>>()
+    private val writeChannelClicks = Channel<List<Click>>()
+    private val writeChannelPurchases = Channel<List<Purchase>>()
+
+    private val uploadChannel = Channel<String>()
+
+
+    fun setup(
+        context: Context,
+    ) {
+        initialize(context)
+
+        launchWriteLoop(writeChannelImpressions, KEY_IMPRESSION_EVENTS)
+        launchWriteLoop(writeChannelClicks, KEY_CLICK_EVENTS)
+        launchWriteLoop(writeChannelPurchases, KEY_PURCHASE_EVENTS)
+
+        launchUploadLoop()
+    }
+
+    suspend fun storeImpression(
+        impressionEvent: ImpressionEvent
+    ) {
+        writeChannelImpressions.send(impressionEvent.impressions)
+    }
+
+    suspend fun storeClick(
+        clickEvent: ClickEvent
+    ) {
+        writeChannelClicks.send(clickEvent.clicks)
+    }
+
+    suspend fun storePurchase(
+        purchaseEvent: PurchaseEvent
+    ) {
+        writeChannelPurchases.send(purchaseEvent.purchases)
+    }
+
+    fun upload() {
+        uploadChannel.trySend("UPLOAD")
+    }
+
+    private fun initialize(context: Context) {
+        applicationContext = context.applicationContext
+    }
+
+    private fun launchUploadLoop() = scope.launch(dispatcher) {
+        for (a in uploadChannel) {
+            val aggregated = aggregateEvents()
+
+            // Actually send
+            // TopsortAnalyticsHttpService.service.reportAggregated(aggregated)
+            println("uploading: $aggregated")
+
+            clear()
+        }
+    }
+
+    private fun <T : JsonSerializable> launchWriteLoop(
+        channel: Channel<List<T>>,
+        key: Preferences.Key<String>
+    ) =
+        scope.launch(dispatcher) {
+            for (events in channel) {
+                val json = StringBuilder()
+                for (event in events) {
+                    json.append(event.toJsonObject().toString())
+                    json.append(",")
+                }
+
+                applicationContext.eventDatastore.edit { store ->
+                    if (store.contains(key)) {
+                        store[key] = store[key] + json.toString()
+                    } else {
+                        store[key] = json.toString()
+                    }
+                }
+            }
+        }
+
+    private suspend fun aggregateEvents(): String {
+        val data = applicationContext.eventDatastore.data.first()
+        val impressions = data[KEY_IMPRESSION_EVENTS]?.trim(',')
+        val clicks = data[KEY_CLICK_EVENTS]?.trim(',')
+        val purchases = data[KEY_PURCHASE_EVENTS]?.trim(',')
+
+        val aggregated = StringBuilder()
```

Comment:
Can we just build and object that aggregates the individual events and serialize it? Manually creating JSONs is very error-prone.